pkg/workloads/cortex/serve/batch.py

```diff
@@ -0,0 +1,343 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import os
+import argparse
+import inspect
+import time
+import json
+import msgpack
+from concurrent.futures import ThreadPoolExecutor
+import threading
+import math
+import asyncio
+from typing import Any
+
+import boto3
+import botocore
+
+from cortex import consts
+from cortex.lib import util
+from cortex.lib.type import API, get_spec
+from cortex.lib.log import cx_logger
+from cortex.lib.storage import S3, LocalStorage, FileLock
+from cortex.lib.exceptions import UserRuntimeException
+
+API_LIVENESS_UPDATE_PERIOD = 5  # seconds
+INITIAL_MESSAGE_VISIBILITY = 120  # seconds
+MESSAGE_RENEWAL_PERIOD = 60  # seconds
+
+local_cache = {
+    "api": None,
+    "job": None,
+    "provider": None,
+    "predictor_impl": None,
+    "predict_route": None,
+    "client": None,
+    "class_set": set(),
+    "sqs": None,
+}
+
+
+class PeriodicGeneratorRunner:
+    def __init__(self, interval, func, *args, **kwargs):
+        self.interval = interval
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+        self.generator = self.func(*self.args, **self.kwargs)
+
+    def start(self):
+        self.timer = threading.Timer(self.interval, self.start)
+        self.timer.start()
+        next(self.generator)
+
+    def stop(self):
+        self.timer.cancel()
+
+
+def dimensions():
+    return [
+        {"Name": "APIName", "Value": local_cache["api"].name},
+        {"Name": "JobID", "Value": local_cache["job"]["job_id"]},
+    ]
+
+
+def success_counter_metric():
+    return {"MetricName": "Succeeded", "Dimensions": dimensions(), "Unit": "Count", "Value": 1}
+
+
+def failed_counter_metric():
+    return {"MetricName": "Failed", "Dimensions": dimensions(), "Unit": "Count", "Value": 1}
+
+
+def time_per_batch_metric(total_time_seconds):
+    return {"MetricName": "TimePerBatch", "Dimensions": dimensions(), "Value": total_time_seconds}
+
+
+def update_api_liveness():
+    threading.Timer(API_LIVENESS_UPDATE_PERIOD, update_api_liveness).start()
+    with open("/mnt/workspace/api_liveness.txt", "w") as f:
+        f.write(str(math.ceil(time.time())))
+
+
+def startup():
+    open("/mnt/workspace/api_readiness.txt", "a").close()
+    update_api_liveness()
+
+
+def build_predict_args(payload, batch_id):
+    args = {}
+
+    if "payload" in local_cache["predict_fn_args"]:
+        args["payload"] = payload
+    if "headers" in local_cache["predict_fn_args"]:
+        args["headers"] = None
+    if "query_params" in local_cache["predict_fn_args"]:
+        args["query_params"] = None
+    if "batch_id" in local_cache["predict_fn_args"]:
+        args["batch_id"] = batch_id
+    return args
+
+
+def renew_message_visibility(queue_url, receipt_handle, initial_offset, interval, *args, **kwargs):
```

Comment:
Double check that this runs for the first time after 60 seconds

Reply:
Removed